src/state_circuit/state.rs

```diff
@@ -0,0 +1,896 @@
+use crate::gadget::{
+    is_zero::{IsZeroChip, IsZeroConfig, IsZeroInstruction},
+    monotone::{MonotoneChip, MonotoneConfig},
+    Variable,
+};
+use halo2::{
+    circuit::{Layouter, Region},
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, Fixed, Selector},
+    poly::Rotation,
+};
+use pasta_curves::arithmetic::FieldExt;
+
+use std::marker::PhantomData;
+
+/*
+Example memory table:
+| address | global_counter | value | flag |
+---------------------------------
+|    0    |   0  |   0   |   1  |
+|    0    |  12  |  12   |   1  |
+|    0    |  24  |  12   |   0  |
+|    1    |   0  |   0   |   1  |
+|    1    |  17  |  32   |   1  |
+|    1    |  89  |  32   |   0  |
+*/
+
+/*
+Example bus mapping:
+
+| global_counter | memory_flag | memory_address | memory_value |
+------------------------------------------------------
+|  12  |      1      |        0       |      12      |
+|  17  |      1      |        1       |      32      |
+|  24  |      0      |        0       |      12      |
+|  89  |      0      |        1       |      32      |
+*/
+
+/// In the state proof, memory operations are ordered first by address, and then by global_counter.
+/// Memory is initialised at 0 for each new address.
+/// Memory is a word-addressed byte array, i.e. a mapping from a 253-bit word -> u8.
```

Comment:
I thought memory in evm was byte addressable?

Reply:
I think that when you use the compressed form the address will be a253 number its just the random linear combination of the 8 bit words making up the 256 bit address.

Reply:
If we can assume a circuit memory size hard bound (say `2**24`), it would be easier to use actual value instead of random linear combination for memory index.

For example, when one is going to do like `mload(250)`, we will need to do memory bus mapping lookup in range (`250`, `250 + 1`, `250 + 2`, ..., `281`) from byte to byte. If we use random linear combination, then some of indexes would have the first byte which exceeds 256 and produces carry to the second byte, which makes the random linear combination harder to calculate (hard do that without requiring intermediate witness).

If we assume a hard bound, we just take first 3 bytes and recompose it to `2**24` value to lookup memory, which seems to be easier, but is incompatible with current evm which doesn't has any hard bound.

Reply:
Yes this makes sense. Hard bound means we can't overflow 2**253 memory elements. 

I was thinking if we used compressed form we would be able to avoid a decompression of the index when it comes from the stack. But you have to decompress to get the ordering right anyway so probably good to use the decompressed form.

Reply:
253 bit width range come from Pasta Curves ?

Reply:
It comes from the bn254 curve we are using. We replace Pasta curve with bn254 and halo2  polynomial commitment with kate polynomial commitment.

Reply:
you mean code "pasta_curves::arithmetic::FieldExt;" has been with bn254 or happens on other branch ?

Reply:
I think it happened on another branch but that is our plan.

Reply:
The BN254 / KZG fork of halo2 is on @kilic's branch: https://github.com/kilic/halo2/tree/kzg

Reply:
@therealyingtong  I see & Thx